### ISSUE TITLE
[HOLD #11795][$1000] No transition animation on timezone selector

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#24139]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. go to settings > profile > timezone
2. now open timezone selector
## Expected Result: 
there should be a transition animation while it opens
## Actual Result: 
no transition animation happens
## Workaround:
Can the user still use Expensify without this being fixed? Have you informed them of the workaround?

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [x] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.3.50-0
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation

https://github.com/Expensify/App/assets/43996225/c23a5293-d20f-4e4e-a8b3-281486fbec95

https://github.com/Expensify/App/assets/43996225/10890fc4-9439-46da-b12c-8f60ef9f081e

**Expensify/Expensify Issue URL:**
**Issue reported by:** @chiragxarora 
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1690882305493399

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~0177d92223610217d2</li>
        <li>Upwork Job ID: 1687469082823692288</li>
        <li>Last Price Increase: 2023-08-11</li>
<li>Automatic offers: </li>
<ul>
<li>chiragxarora | Contributor | 26163977</li>
<li>chiragxarora | Reporter | 26163980</li>
</ul></li>
    </ul>
</details>

**Comment by kavimuru** [comment]
## Proposal from @chiragxarora 
### Please re-state the problem that we are trying to solve in this issue.
 no transition anmation on timezone selector
### What is the root cause of that problem?
Root cause of the issue is focusing logic in our `SelectionListRadio` component, as we can see currently timezone page uses a slection radio list.
Now the current logic of focusing the text input here is defined like this
https://github.com/Expensify/App/blob/52b8ddc2a5edc1b2d2ddbde20b95779dbbdb6380/src/components/SelectionListRadio/BaseSelectionListRadio.js#L173-L188
Currently the text input is focused conditionally based on `shouldDelayFocus` prop and we have not used that prop in our timezone selector, so it instantly focuses the input leading to no animation
https://github.com/Expensify/App/blob/52b8ddc2a5edc1b2d2ddbde20b95779dbbdb6380/src/pages/settings/Profile/TimezoneSelectPage.js#L119-L126
### What changes do you think we should make in order to solve the problem?
We need to alter the focusing logic in that page and we shall wait for the transitions to end before we focus. For this, we need to wrap our component in ScreenWrapper and then we can use the `onEntryTransitionEnd` prop which is called after transitions end, and focus the textinput inside it using the ref we already have in the page.
Edit: Upon further investigation, I found out that some radio lists are rendered inside the modal like year picker, state picker, country picker, and for them, onEntryTransitionEnd is never called, so to make it for all the radio lists, we can add a check state to look for transition end (for the non modal cases) and focus them accordingly
```js
const [didTrasitionEnd, setDidTransitionEnd] = useState(false);
onEntryTransitionEnd={() => setDidTransitionEnd(true)}
```
and replace this line
https://github.com/Expensify/App/blob/52b8ddc2a5edc1b2d2ddbde20b95779dbbdb6380/src/components/SelectionListRadio/BaseSelectionListRadio.js#L178-L178
with this
```js
if(didTrasitionEnd) textInputRef.current.focus();
```
<details>
  <summary>Results</summary>
https://github.com/chiragxarora/pdf-download/assets/54641656/8f21b0f8-d3c4-45d8-ba00-2417015fe04d
</details>
### What alternative solutions did you explore? (Optional)
We can keep the SelectionListRadio component as it is and simply add shouldDelayFocus prop to the selector in timezone
 page

**Comment by jfquevedol2198** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
No transition animation on timezone selector

### What is the root cause of that problem?
https://github.com/Expensify/App/blob/52b8ddc2a5edc1b2d2ddbde20b95779dbbdb6380/src/components/SelectionListRadio/BaseSelectionListRadio.js#L173-L180
Without `shouldDelayFocus` props, `BaseSelectionListRadio` component with TextInput element tries to focus on TextInput when its mounted. Focusing on Text in the beginning prevents transition of ScreenWrapper component.

### What changes do you think we should make in order to solve the problem?
Pass `shouldDelayFocus` props in `SelectionListRadio` element inside `TimezoneSelectPage` render method.
```javascript
<SelectionListRadio
    shouldDelayFocus
    ...
/>
```

### What alternative solutions did you explore? (Optional)
N/A

**Comment by b4s36t4** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
No transition animation on timezone selector
### What is the root cause of that problem?

RCA for this issue pretty straight forwarded and already observed at many places. Same here in the timezone select page we have an input which is getting autoFocused which causing the cluttered/no-animation for component.

Since we're setting focus to the input before completing the rendering of modal(Transition) this issue happening.

### What changes do you think we should make in order to solve the problem?

I feel this proposal is tricky and long but I feel this should be done like this only.

Steps to solve this.

Step 1:
https://github.com/Expensify/App/blob/9294e9eacd53ba2ba805c6e33f18c796ccc3306d/src/components/SelectionListRadio/index.js#L29 here we're passing the `ref` to `BaseSelectionListRadio` component which is functional component. 

But we're not using `forwardRef` here which is causing an error in console. As part of the solution we have to fix this issue which will let us use the ref inside the `BaseSelectionListRadio` component.

Step 2:
Write an imperative handler to focus the input. Since `BaseSelectionListRadio` is a functional component, using `useImperativeHandler` hook should help here.

Inside the hook we should have a function maybe like `focusInput` which focuses the input.

```jsx
    useImperativeHandlers(props.forwardedRef, () => ({
        focusInput: () =>{
            textInputRef.current.focus()
        }
    }))

```
Step 3:
Create a new prop called `autoFocus` and set it's default value to `true` to support the backward compatibility this default value help us to not get regressions and not to raise the bugs where the same component get's used.

Step 4:
Restrict focusing input when the `autoFocus` prop is false https://github.com/Expensify/App/blob/ec7b26703980e1cb645a382f68a2f743e59f1520/src/components/SelectionListRadio/BaseSelectionListRadio.js#L174

```js
     if (shouldShowTextInput && props.autoFocus) {
        // Run the focus logic here.
        // we also have to add `props.autoFocus` to `useEffect` dependency.
     }
```


which results makes the animation to work perfectly when this false passed here https://github.com/Expensify/App/blob/9294e9eacd53ba2ba805c6e33f18c796ccc3306d/src/pages/settings/Profile/TimezoneSelectPage.js#L119 in this component.

```jsx
<SelectionListRadio
    {...props}
    autoFocus={false}
    />
```

Step 5:
use `onEntryTransitionEnd` prop for the `ScreenWrapper` component and access the ref and call `focusInput` function which we created using the imperative handler.

> create a new ref inside `TimezoneSelectPage` component. `ex: this.selectionListRef = createRef(null)`

> Assign the `selectionListRadio`'s ref to `this.selectionListRef`.

```jsx
 ref={(el) => this.selectionListRef.current = ref}
```

call the `focusInput` inside the `onEntryTransitionEnd` function.

```jsx
onEntryTransitionEnd={() => {
    this.selectionListRef.current && this.selectionListRef.current.focusInput()
}}
```



### What alternative solutions did you explore? (Optional)
Simply pass the `shouldDelayFocus` here should also do the work. But previously I proposed same type of solution like adding `shouldDelayFocus` got rejected because `shouldDelayFocus` is using a time-out function, so proposing the above solution.

Or if we feel all this is complex logic which might cause regressions later down the line we could simply do conditional rendering inside `TimezoneSelectPage` component.

We could simple use `didScreenTransitionEnd` passed by `ScreenWrapper` component and show the loader while it was false and show the component when it becomes `true` which does the same work.

> The proposed solution is also working with the focusing even after re-fresh whereas other solutions might work or not.

## Result

https://github.com/Expensify/App/assets/59088937/40f61c37-88e0-439c-b708-b0099bf6324f

**Comment by robertKozik** [comment]
Hi all! Thank you all for your proposals!
1. @chiragxarora - I like your proposal, could you provide me where we are displaying the radio list on modal? I want to check this scenario too
2. @jfquevedol2198 - I believe we should't accept the proposal which enables the `useTimeout` workaround when we have other possible solutions. `useTimeout` is the last resort

**Comment by b4s36t4** [comment]
@robertKozik handling conditional logic inside the base component doesn't make sense to me. I feel since we're having a `ref`(but not handled) ref can handle this here. And this is the approach at many places we're following i.e accessing the ref to manage the focus if not going with `shouldDelayFocus` prop. Any thoguhts?

**Comment by robertKozik** [comment]
I have mixed thoughts for this. The @chiragxarora approach makes sense for me, as it's quite simple and doing it's job. But you pointed out good counter argument. It's better to keep base component with no "outside" logic

@b4s36t4 did you check these places which @chiragxarora pointed out - with timezone selector as modal? Does your solution work for them too?

**Comment by b4s36t4** [comment]
@robertKozik I think the places which @chiragxarora mentioned are component not screens. As far as I know screens only have animations not components. Although I'll get back to you once I check them.

**Comment by b4s36t4** [comment]
@chiragxarora there are no drawbacks with your issue. the solution which is proposed is having outside logic which is the state you'll be introducing `didTrasitionEnd` and you're passing that state to the entire base component. Also to handle the focus you need to add that to dependencies of `useEffect` which might (not sure) can cause regressions.

The main thing I'm worried is passing a state which is derived from a screen to the base component is not what the solution should do.

I feel we have `forwardRef` in the `SelectionListRadio` component which is passed down to `BaseSelectionListRadio` and using imperative handle does make sense to me. 

Also the modals which you listed above uses `shouldDelayFocus` prop and they're working prefect. I'm not against using `shouldDelayFocus` if @robertKozik feels ok with `shouldDelayFocus` then I feel that's ok too. All I'm worried about brining the new state all the way to base component which is not good.

Also it doesn't have to be `onEntryTransitionEnd` to focus the input, for modals we have `onModalHide` which will only be called after modal is rendered (checked ;)).

Decision is upto @robertKozik. I feel we don't touch the ones which are working fine with `shouldDelayFocus` or let's just go with `shouldDelayFocus`.

**Comment by robertKozik** [comment]
@chiragxarora The only downside of your proposal is that we are sticking to the forcing focus on textInput as even though expected for now, can be trouble-some for feature possible use-cases (f.e when we would have this component and another textInput rendered). Another one, but more implementation-specific, is that we maybe don't want to wrap entire RadioList inside ScreenWrapper as it provides a lot more than transition callback prop, but rather stick to the implement separate listener for thet event (but this can be discussed after proposal selection if needed).

To sum up, **I'll come back tomorrow** to this issue, once more  go through this conversation and **choose the proposal**.

Thank you both for this constructive conversation. 🙇🏼‍♂️

**Comment by robertKozik** [comment]
Sure! I'm always open for discussion!

**Comment by robertKozik** [comment]
Sorry for keep you waiting! 🙇🏼 
I one more checked the discussion thread here and I think I'm leaning towards @chiragxarora approach. Hoisting the focus logic outside the component is not necessary in the current use-cases of the radio lists. Although it can become relevant in the feature we cannot be certain about this, so for today it's more like over-engineering which can be more prone to regressions.

I would go with the @chiragxarora [proposal](https://github.com/Expensify/App/issues/24139#issuecomment-1673480130)

🎀 👀 🎀 C+ reviewed

**Comment by dangrous** [comment]
Hi! Looking back through this, a couple of questions:
* I'd love to sync up on `shouldDelayFocus` - I know we try to avoid using timeouts, but this isn't exactly adding a new one, just using an existing feature in the `SelectionListRadio` component. It seems like that would be simplest? But I haven't really dealt with much that would need it before.
* Do we want to expand this issue to include the other selectors that don't currently have transitions (e.g. I noticed the currency selector doesn't seem to)

**Comment by b4s36t4** [comment]
@dangrous `shouldDelayFocus` should do the work. Most of the time proposal get's rejected if it includes this, so people tend to follow alternate approaches.

**Comment by dangrous** [comment]
Ah got it. Let me do some digging to make sure I understand why we would typically reject those, so I don't do something I shouldn't!

**Comment by b4s36t4** [comment]
@dangrous only reason `shouldDelayFocus` is being rejected is that it uses `setTimeoout`.

**Comment by b4s36t4** [comment]
https://github.com/Expensify/App/issues/22271#issuecomment-1629784328 The proposal got rejected because of the same reason here.

**Comment by dangrous** [comment]
Okay thanks all! @chiragxarora's proposal makes sense and tracks with the way we're looking to standardize [here](https://expensify.slack.com/archives/C01GTK53T8Q/p1690301750089799?thread_ts=1689957344.255689&cid=C01GTK53T8Q). I'll assign you!

**Comment by dangrous** [comment]
Based on [Slack](https://expensify.slack.com/archives/C01GTK53T8Q/p1692307751392999), I believe we have decided to close out this PR - but not the issue - and it will be fixed in the refactor of this component. We should still pay out @robertKozik and @chiragxarora as if the PR went through. Does that sound right to everyone?

**Comment by robertKozik** [comment]
@dangrous Sure! I'd like to remind that I'm from Software Mansion (an expert agency), so I'm not eligible for payment.

**Comment by dangrous** [comment]
Okay cool, I'll put this on hold for https://github.com/Expensify/App/pull/22622 for now, and we can test and confirm and pay once that's all complete

**Comment by dangrous** [comment]
Thanks @b4s36t4! I've called out in that issue that it'll likely get solved by the SelectionList refactor just like this one.

**Comment by dangrous** [comment]
Still on hold

**Comment by b4s36t4** [comment]
@dangrous I believe SelectionList refactor is complete, isn't it?

**Comment by thiagobrez** [comment]
Hey @dangrous! Yep, phase 3 is ongoing, and this was solved in phase 2.

More specifically, the regression was solved by this PR https://github.com/Expensify/App/pull/25921.

And the improvements to the logic we discussed in the slack thread was solved here (on hold because
of merge freeze)
https://github.com/Expensify/App/pull/26415

**Comment by dangrous** [comment]
Oh, sweet! @robertKozik do you mind testing this again to see if we can close out?

**Comment by dangrous** [comment]
bump for testing @robertKozik when you have time! Should be quick - thanks

**Comment by robertKozik** [comment]
Sure! Missed that one, on it 👀

**Comment by robertKozik** [comment]
Can confirm, transition is present and issue is fixed 🚀

**Comment by dangrous** [comment]
Sweet! So I think we just need payment for @chiragxarora, as both reporter and contributor - even though the PR was not used, we [agreed](https://github.com/Expensify/App/issues/24139#issuecomment-1686937761) to pay as if it had. Does that seem right to everyone?

**Comment by bfitzexpensify** [comment]
All sorted!
